zebrad/src/commands/start/sync.rs

```diff
@@ -392,22 +446,37 @@ where
             let span = tracing::info_span!("block_fetch_verify", ?hash);
             let mut verifier = self.verifier.clone();
             let task = tokio::spawn(
+                // TODO: refactor this task into its own async function, and
+                //       wrap_err the result with hash
                 async move {
-                    let block = match block_req.await {
-                        Ok(zn::Response::Blocks(blocks)) => blocks
-                            .into_iter()
-                            .next()
-                            .expect("successful response has the block in it"),
-                        Ok(_) => unreachable!("wrong response to block request"),
-                        Err(e) => return Err(e),
-                    };
-                    metrics::counter!("sync.downloaded_blocks", 1);
-
-                    let result = verifier.ready_and().await?.call(block).await;
-                    metrics::counter!("sync.verified_blocks", 1);
-                    result
-                }
-                .instrument(span),
+                    async {
+                        let block = match block_req.await {
+                            Ok(zn::Response::Blocks(blocks)) => blocks
+                                .into_iter()
+                                .next()
+                                .expect("successful response has the block in it"),
+                            Ok(_) => unreachable!("wrong response to block request"),
+                            // Make sure we can distinguish download and verify timeouts
+                            Err(e) => Err(eyre!(e)).wrap_err("failed to download block")?,
+                        };
+                        metrics::counter!("sync.downloaded.block.count", 1);
+
+                        let result = verifier
+                            .ready_and()
+                            .await
+                            .map_err(|e| eyre!(e))
+                            .wrap_err("verifier service failed to be ready")?
+                            .call(block)
+                            .await
+                            .map_err(|e| eyre!(e))
+                            .wrap_err("failed to verify block")?;
+                        metrics::counter!("sync.verified.block.count", 1);
+                        Result::<block::Hash, Report>::Ok(result)
+                    }
+                    .instrument(span)
+                    .await
+                    .map_err(|e| (e, hash))
```

Comment:
@yaahc I tried using another `wrap_err(hash)` here, but it turned out to be really complicated and error-prone.

The downcasting to a hash was difficult to get right, particularly with the intervening `instrument`.

If you have any suggestions for a better way to pass back the hash, that would be great!

Reply:
For cases where you're capturing the span to log it I'd remove them because the `hash` is already included in the report via `span`. When it constructs the report it should capture the spantrace including `span`. If you're not seeing that output its because the span is set to `info` and your daemon isn't setup to capture info spans, so I'd move that to `warn` or `error` if its important to capture it in all error reports.

As for the place where you're checking if the hash is contained in the state. I think your approach of bundling it on the outside is probably the best one here. In theory we could do some shenanigans with `bundlerr` that basically does the same thing `eyre`'s downcast magic does, but its a mess and hacky and we'd probably have to vendor `color-eyre` to do it properly so we can control our own `EyreReport` impl.

Reply:
This PR just uses the hash as a way to ignore some errors, so it seems like the current approach is good.